### PR TITLE
replace for strings with multiple patterns

### PR DESCRIPTION
**Pull request by KlausC**
This PR supersedes https://github.com/JuliaLang/julia/pull/25396.

Following improvements were made, following the critics in the previous version.

1. Performance in terms of time and memory allocation has been drastically improved (O(n)).
2. Performance is comparable or better to the old one-pair version
3. Code is type stable
4. Memory allocation is O(1) plus the allocation in user provided input or output functions
5. Uses generated function to exploit type information about  patterns and replacements when run first time with those type combinations.
6. Code was extracted into new base/string/replace.jl overcome build dependencies.

I would like to ask @nalimilan  and @stevengj to have a look on this, again.

### CODE REVIEW

File: base/strings/replace.jl [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+
+# types input patterns are converted to
+const TF = Union{Regex,String,Callable}
+# types output patterns are converted to
+const TS = Union{Char,String,Callable,SubstitutionString}
+
+# from reflection.jl::hasmethod
+function type_hasmethod(@nospecialize(ft::DataType), @nospecialize(t), world=typemax(UInt))
+    t = Tuple{ft, t...}
+    ccall(:jl_method_exists, Cint, (Any, Any, UInt), ft.name.mt, t, world) != 0
```
Comment (vtjnash) [inline base/strings/replace.jl]: It's illegal to call this from a generated function: it's simply never permissible to introspect the method tables.

**Review by vtjnash — CHANGES_REQUESTED** [review]
Seems like this should be in a stdlib package – it seems like rather a lot of code to move into base.

**Comment by KlausC** [comment]
What is the rationale behind the term "illegal"?
I need some technical hints:

1. Would it be sufficient to move the calling of `type_hasmethod`outside the generated function?
2. Is there legal way to find out, if a certain method signature exists, given only the type of the function?
3. Would there be a chance to put the function `type_hasmethod` into reflection.jl?
4. If we move the implementation into a stdlib package, should it be a new one `Replace` and would it be legal to extend the Base.replace function in there?

BTW: the implemented approach works as expected and leads to extreme performance improvements.
For example: calling `uppercase('a')` is  6666 times faster than `uppercase("a")`.
So my problem is not to do illegal business and harm the rules, but to exploit this performance advantage in a generic way.

**Comment by nalimilan** [comment]
I think by "illegal" @vtjnash means that assumptions the compiler makes are not respected, which means many wrong things can happen.

Thanks for exploring this, but TBH that's a lot of new complex code so you may have a hard time finding somebody able to review this given that people are focused on finishing things that cannot be added after 1.0. Also the ratio complexity/benefit isn't very high so I'm not sure it will be accepted in Base. Maybe we should start with a simpler API which can be implemented efficiently with less efforts.

**Comment by vtjnash** [comment]
> Is there legal way to find out, if a certain method signature exists, given only the type of the function?

It's perfectly legal to find out if it exists, that's just dispatch. It's not legal for generated functions to compute that information and return it. Outside of a generated function, you have the values, so you can just call the function. It's also currently slow, and while we could optimize it, we have so far avoided ever defining an API that depended on what methods are defined. Thus far, the preference has been for requiring explicit traits or not dispatching the behavior based on types at all.

Implementing this in a straightforward, readable manner (e.g. without generated functions) would go far towards getting this into a state where we could review and merge this.

**Comment by KlausC** [comment]
> Implementing this in a straightforward, readable manner (e.g. without generated functions) would go far towards getting this into a state where we could review and merge this.

This PR supersedes #25396. That was a first attempt to solve the problem in a straightforward, readable manner. Actually, it seemed to me a suitable generalization of the existing API. The concept of the algorithm was easy to generalize, and I was happy to present a corresponding implementation. (It was straightforward, readable, and not more than 40 lines of code).
This approach was bashed, mainly because it was not as performing as well as a non-generic  hand-written code.

In the meanwhile, I learned, that there is a trade-off between readability of code and performance, also in Julia.
I tried hard to save the initial approach and learnt how to perform coverage-, runtime- and allocation analysis, type stability and `@code_warntype`.
The core problem, which I could not resolve without using meta-programming, was that the call of a variable set of functions within the inner loop of the algorithm. After removing all type-instabilities, it each function still used 16 or 32 bytes of memory allocation. That bloated allocated memory and runtime. I was not able to find out, how the compiler could help me. I would really have preferred to pull the dispatch work for these call outside of the inner loop, but did not find.

So it seemed natural to me to have a look at `@generated function`s. The idea was now to generate code, that looks like hand-coded, without method invocations within the inner loop. It was a pain to write this code, as it is to read it. Nevertheless it was well structured, but complex. The end result, after investing a lot of work, was excellent in terms of performance: for long input data, it outperforms the existing code for a single pattern-replacement pair. In the simple cases, when hard-coded specialisations are available, the performance is not worse than 3 times the hard-coded, which is acceptable.

```
julia> using BenchmarkTools
julia> n = 10^6; s = "a"^n * "b";
julia> @btime replace(s, islower=>uppercase, ""=>""); # using multiple-pairs
  49.920 ms (27 allocations: 1.15 MiB)
julia> @btime replace(s, islower=>uppercase); # using the original implementation
  247.972 ms (7998992 allocations: 306.31 MiB)

julia> @btime replace(s, 'x'=>'A', 'b'=>'B', islower=>uppercase);
  61.400 ms (48 allocations: 1.15 MiB)
@btime replace_sjhand(s); # corresponding hard-coded
  29.030 ms (20 allocations: 1.00 MiB)
```

@vtjnash, @nalimilan, @stevengj, @StefanKarpinski 
Maybe you understand, that I am now a little bit disappointed, frustrated, discouraged because all the effort leads to nothing.
I think, I will draw the following conclusions:
1. If it is possible for me to write ugly code, which is performing efficiently, by using generated functions, that should, in theory be also possible for the compiler.
2. The method invocations in inner loops are preventing optimal performance.
3.  Maybe a reason for that is, the compiler only uses the abstract type of those functions (`Function`), not the real type (`typeof(uppercase)`) for code generation.
4. I could imagine, that a method had a type or supertype, which defines signature and return type, but abstracts from the function name; the compiler could make use of this information (if specified by the user as argument type) and generate code, which dispatches by the method name only. Actually that can be done by a user-defined `struct`, but the compiler cannot exploit it.

I really understand, that most of you are busy with more important subjects. I would like to help you as far as I can, and not produce inefficient extra work for you. Maybe some of you have a hint for me, in which direction I should develop.

**Comment by stevengj** [comment]
Sometimes it is necessary to write really complicated, tricky code, to get high performance and very general functionality.   But if it's not core functionality of the language (e.g. not something that we need for other parts of Base), it's a reasonable thing to put that complexity into a separate package.

**Comment by nalimilan** [comment]
> If it is possible for me to write ugly code, which is performing efficiently, by using generated functions, that should, in theory be also possible for the compiler.
>     The method invocations in inner loops are preventing optimal performance.
>     Maybe a reason for that is, the compiler only uses the abstract type of those functions (Function), not the real type (typeof(uppercase)) for code generation.

AFAIK this shouldn't be the case. Methods are always specialized on the type of function arguments when such functions are called directly (instead of just being passed to another function). You can always force this too by using a type parameter `T<:Function`.

Would you be able to put together a simple example where it's not happening? Though IIRC from my experiments with implementing `broadcast` for `NullableArray` sometimes it's splatting which can incur a penalty. Maybe an interesting test would be to check whether a one-argument version (without varargs) can be made efficient without generated functions.

Sorry that it takes so much time without a clear perspective of being included in Base. That's why I suggested we could try to find a subset of the API which is easy to implement efficiently for now.

**Comment by KlausC** [comment]
@nalimilan 

> AFAIK this shouldn't be the case. Methods are always specialized on the type of function arguments when such functions are called directly (instead of just being passed to another function). You can always force this too by using a type parameter T<:Function.
Would you be able to put together a simple example where it's not happening?

Please have a look at [kctest.jl](https://github.com/KlausC/Replace.jl/blob/master/src/kctest.jl) and [kctest.jl.mem](https://github.com/KlausC/Replace.jl/blob/master/src/kctest.jl.mem) functions `hard` and `soft/soft2` to see the difference between hard-coded method calls and the same methods given as vararg or vector list. Between varargs and vectors, there seems no difference. I did not yet look for the difference between single function argument and multiple function arguments.

**Comment by nalimilan** [comment]
OK, too bad. Instead of a vector, I've also tried with a tuple, and it gives the same result. It's disappointing because in CategoricalArrays' `recode`, iteration over a tuple of value is unrolled, but these are not functions so it's a bit different.

**Pull request closed without merging**